scylla/src/transport/topology.rs

```diff
@@ -408,55 +420,88 @@ async fn query_peers(conn: &Connection, connect_port: u16) -> Result<Vec<Peer>,
         "system.local query response was not Rows",
     ))?;
 
-    let mut result: Vec<Peer> = Vec::with_capacity(peers_rows.len() + 1);
-
     let typed_peers_rows =
         peers_rows.into_typed::<(IpAddr, Option<String>, Option<String>, Option<Vec<String>>)>();
 
-    // For the local node we should use connection's address instead of rpc_address unless SNI is enabled (TODO)
-    // Replace address in local_rows with connection's address
-    let local_address: IpAddr = conn.get_connect_address().ip();
-    let typed_local_rows = local_rows
-        .into_typed::<(IpAddr, Option<String>, Option<String>, Option<Vec<String>>)>()
-        .map(|res| res.map(|(_addr, dc, rack, tokens)| (local_address, dc, rack, tokens)));
-
-    for row in typed_peers_rows.chain(typed_local_rows) {
-        let (ip_address, datacenter, rack, tokens) = row.map_err(|_| {
-            QueryError::ProtocolError("system.peers or system.local has invalid column type")
-        })?;
-
-        let tokens_str: Vec<String> = tokens.unwrap_or_default();
+    let local_ip: IpAddr = conn.get_connect_address().ip();
+    let local_address = SocketAddr::new(local_ip, connect_port);
+
+    let typed_local_rows =
+        local_rows.into_typed::<(IpAddr, Option<String>, Option<String>, Option<Vec<String>>)>();
+
+    let untranslated_rows = typed_peers_rows
+        .map(|res| res.map(|peer_row| (false, peer_row)))
+        .chain(typed_local_rows.map(|res| res.map(|local_row| (true, local_row))));
+
+    let translated_peers_futures = untranslated_rows.map(|untranslated_row| async {
+        match untranslated_row {
```

Comment:
You can use `?` operator in closures and futures which return `Result`. You can reduce nesting by rewriting the first match as:

```rust
let (is_local, (untranslated_ip_addr, datacenter, rack, tokens)) = untranslated_row.map_err(|_| {
    QueryError::ProtocolError("system.peers or system.local has invalid column type")
})?;

// Rest of the code for the `Ok(_)` case
```

The resulting code will be flatter and the diff will become smaller.

However, `rustc` will need your help in order to determine the return type of the `async` block because it won't be able to infer the `E` in `Result<T, E>`. It won't be able to do it from the `?` expression because the `?` operator is allowed to convert the error type with an `Into` conversion, and you won't be returning any `Err(x)` variants where `x` would be a known type.

I don't think it is possible to annotate the return type of an async block directly - however, you can use the turbofish operator to specify `E` when returning the `Ok` variant, e.g.:

```rust
return Result::<_, QueryError>::Ok(None);
```